### PR TITLE
Reflection dockbloc 3

### PR DESCRIPTION
**Pull request by theofidry**
Follow up of #263:

- Adapted the code to be able to run in both v3 and v2 of `phpdocumentor/reflection-docblock`
- Added a Travis build with `phpdocumentor/reflection-docblock 3.x`

cc @dunglas.

### CODE REVIEW

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 1]
```diff
@@ -79,5 +102,41 @@ public function getPriority()
...
+
+    /**
+     * @param \ReflectionClass $reflectionClass
+     *
+     * @return LegacyMethodTag[]|Method[]
+     */
+    private function getClassTagList(\ReflectionClass $reflectionClass)
+    {
+        try {
+            $phpdoc = (null === $this->docBlockFactory || null === $this->contextFactory)
```
Comment (aik099) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: That doesn't look like proposed adapter creation. This IF statement checking which Reflection library version is being used is called each time, which is bad performance wise.

Comment (dunglas) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: A `if` statement is nothing compared to all parsing operations involved...

**Comment by theofidry** [comment]
@aik099 I went with the simplest solution at hand. But if you prefer I can add a true adapter to retrieve tags and let you guys choose which one you prefer.

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php (hunk 1 above)
Comment (sroze) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: The main point is that it's not an adapter at all :)

**Comment by theofidry** [comment]
@aik099 here you go

File: src/Prophecy/PhpDocumentor/ClassAndInterfaceTagRetriever.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+use phpDocumentor\Reflection\DocBlock\Tag\MethodTag as LegacyMethodTag;
+use phpDocumentor\Reflection\DocBlock\Tags\Method;
+
+/**
+ * @author Théo FIDRY <theo.fidry@gmail.com>
+ */
+final class ClassAndInterfaceTagRetriever implements MethodTagRetrieverInterface
+{
+    /**
+     * @var MethodTagRetrieverInterface
```
Comment (vincentchalamon) [inline src/Prophecy/PhpDocumentor/ClassAndInterfaceTagRetriever.php]: To be consistent with property `$tagRetriever` on class `MagicCallPatch`, you should add PHPDoc everywhere or nowhere

Comment (theofidry) [inline src/Prophecy/PhpDocumentor/ClassAndInterfaceTagRetriever.php]: yeah this one should be removed

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 3]
```diff
@@ -11,17 +11,27 @@
...
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Théo FIDRY <theo.fidry@gmail.com>
  */
 class MagicCallPatch implements ClassPatchInterface
 {
+    private $tagRetriever;
+
+    public function __construct(MethodTagRetrieverInterface $tagRetriever = null)
+    {
+        $this->tagRetriever = (null === $tagRetriever) ? new ClassAndInterfaceTagRetriever() : $tagRetriever;
```
Comment (vincentchalamon) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: Parentheses are useless

File: src/Prophecy/PhpDocumentor/ClassAndInterfaceTagRetriever.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function __construct(MethodTagRetrieverInterface $classRetriever = null)
+    {
+        if (null !== $classRetriever) {
+            $this->classRetriever = $classRetriever;
+
+            return;
+        }
+
+        $this->classRetriever = (class_exists('phpDocumentor\Reflection\DocBlockFactory') && class_exists('phpDocumentor\Reflection\Types\ContextFactory'))
```
Comment (vincentchalamon) [inline src/Prophecy/PhpDocumentor/ClassAndInterfaceTagRetriever.php]: Same here, parentheses are useless. But as this test is very long, you should write it using `if` syntax

**Comment by dunglas** [comment]
phpDocumentor/TypeResolver#16 has been merged and a new version tagged. Tests should pass now, @theofidry can you add a conflict section in `composer.json` for Type Resolver < 0.1.7?

**Comment by everzet** [comment]
Still failing

**Comment by theofidry** [comment]
@everzet there is something really weird on docblock side so needs to debug it :/ Maybe I'll have time to check it next week

**Comment by everzet** [comment]
I'm closing this PR as it is seems to be abandoned. Please feel free to rebase and resubmit it if that's not the case!

**Comment by iltar** [comment]
As the issue is still valid, any chance this will be re-opened or should a new PR be made? Symfony 3.1 is released and this dependency will prevent people from upgrading.

**Comment by stof** [comment]
Well, if @theofidry wants to finish it, I can reopen it. If you want to take it over to finish it instead, you should fetch the branch from @theofidry's fork and then create a new PR with the updated version.

**Comment by dunglas** [comment]
I'm working on this one right now.

**Comment by GrahamCampbell** [comment]
@dunglas :+1:

**Pull request closed without merging**